src/index_notation/index_notation.cpp

```diff
@@ -2597,6 +2690,241 @@ IndexStmt makeConcreteNotation(IndexStmt stmt) {
+        Reduction reduction = to<Reduction>(rhs);
+        topLevelReductions.push_back(reduction.getVar());
+        rhs = reduction.getExpr();
+      }
+
+      if (rhs != node->rhs) {
+        stmt = Assignment(node->lhs, rhs, Add());
+        if (forallIndexVars.empty()) {
+          for (auto &i : util::reverse(topLevelReductions)) {
+            stmt = forall(i, stmt);
+          }
+        } else {
+          reductionAndFreeVars.insert(reductionAndFreeVars.end(), topLevelReductions.begin(), 
+                                      topLevelReductions.end());
+        }
+      }
+      else {
+        stmt = node;
+      }
+    }
+  };
+  auto rewriter = RemoveTopLevelReductions(forallIndexVars);
+  stmt = rewriter.rewrite(stmt);
+  reductionAndFreeVars = rewriter.reductionAndFreeVars;
+  // This gets the list of indexVars on the rhs of an assignment
+  // TODO: check to make sure that we want to get ALL rhs indexVars (not just the upper level)
+  vector<IndexVar> rhsVars;
+  match(stmt,
+        function<void(const AccessNode*, Matcher*)>([&](const AccessNode* op, Matcher* ctx) {
+          for (auto &i : op->indexVars) {
+            if (std::find(rhsVars.begin(), rhsVars.end(), i) == rhsVars.end()) {
+              rhsVars.push_back(i);
+            }
+          }
+        }),
+        function<void(const AssignmentNode*, Matcher*)>([&](const AssignmentNode* op, Matcher* ctx) {
+          ctx->match(op->rhs);
+        })
+  );
+
+  // Emit the freeVars as foralls if the freeVars are fully derived
+  // else emit the fully derived descendant of the freeVar found in rhsVars
+  if (forallIndexVars.empty()) {
+    for (auto &i : util::reverse(freeVars)) {
+      if (provGraph.isFullyDerived(i))
+        stmt = forall(i, stmt);
+      else {
+        auto derivedVars = provGraph.getFullyDerivedDescendants(i);
+        IndexVar derivedI = *rhsVars.begin();
+        for (auto &derivedVar : derivedVars) {
+          if (std::find(rhsVars.begin(), rhsVars.end(), derivedVar) != rhsVars.end()) {
+            derivedI = derivedVar;
+          }
+        }
+        stmt = forall(derivedI, stmt);
+      }
+    }
+  } else {
+    reductionAndFreeVars.insert(reductionAndFreeVars.end(), freeVars.begin(), freeVars.end());
+    for (auto &i : util::reverse(forallIndexVars)) {
+      if (std::find(reductionAndFreeVars.begin(), reductionAndFreeVars.end(), i) != reductionAndFreeVars.end())
+        stmt = forall(i, stmt);
+      else {
+        auto ancestorVars = provGraph.getUnderivedAncestors(i);
+        IndexVar ancestorI = *reductionAndFreeVars.begin();
+        for (auto &ancestorVar : ancestorVars) {
+          if (std::find(reductionAndFreeVars.begin(), reductionAndFreeVars.end(), ancestorVar) 
+              != reductionAndFreeVars.end()) {
+            stmt = forall(i, stmt);
+          }
+        }
+      }
+    }
+  }
+
+  // Replace other reductions with where and forall statements
+  struct ReplaceReductionsWithWheres : IndexNotationRewriter {
```

Comment:
This looks unchanged from the one in `makeConcreteNotation`...maybe this should be refactored out?

Reply:
Done